### PR TITLE
Add rounded corners for docked mode

### LINKED ISSUE #34: Rounded corners for docked mode?
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/touch-bar-simulator/issues/34)

First of all, thank you for releasing v3.0.0! It makes my wildest dreams come true.

I have one small request - would it be possible to make the sharp corners of the docked mode window style into rounded corners?

Cheers 😄

> There is a $40.00 open bounty on this issue. Add more on [Issuehunt](https://issuehunt.io/r/sindresorhus/touch-bar-simulator/issues/34).
> - Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/touch-bar-simulator/) to discover more funded issues.
> - Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on Issuehunt to raise funds.

### PR DESCRIPTION
**Pull request by benkoska**
Fixes #34 
Set cornerRadius to 5.0.

### CODE REVIEW

**Comment by sindresorhus** [comment]
> but since the docked mode is not using a window frame, we will have to implement the rounded corners for the **bottom left/right** manually.

**Comment by DeveloperBen** [comment]
image

**Comment by sindresorhus** [comment]
It should only be round on the edge that is not touching the screen edge. In your screenshot, it should only have rounded top-left and top-right corners.

**Comment by DeveloperBen** [comment]
Oh, ok I will try to do that

**Comment by GoldenChaos** [comment]
(To inject some subjective opinion/feature creep, it would be great to eventually be able to specify a distance from the edge of the screen and have all 4 corners be rounded if this is true!)

**Comment by DeveloperBen** [comment]
@sindresorhus I updated it
image
image

**Comment by sindresorhus** [comment]
If I set "Dock to Bottom", then restart the app, and then select "Dock to Top", there's some kind of visual glitch:

Screenshot 2019-09-13 at 14 17 09

File: Touch Bar Simulator/TouchBarWindow.swift [hunk 1]
```diff
@@ -295,6 +300,17 @@ final class TouchBarWindow: NSPanel {
 	func removeTitlebar() {
 		styleMask.remove(.titled)
 	}
+	
+	func roundCorners(_ corners: CACornerMask) {
+		self.contentView?.layer?.cornerRadius = 7.5
+		self.contentView?.layer?.maskedCorners = []
+
+		self.contentView?.layer?.maskedCorners = corners
+		self.contentView?.layer?.borderWidth = 0.0
```
Comment (sindresorhus) [inline Touch Bar Simulator/TouchBarWindow.swift]: This logic could be moved out into a `CALayer` extension to be reusable.

Could also add a `NSWindow` extension that used this and also handled `isOpaque`, etc.

File: Touch Bar Simulator/TouchBarWindow.swift [hunk 2]
```diff
@@ -295,6 +300,17 @@ final class TouchBarWindow: NSPanel {
...
+		self.contentView?.layer?.borderColor = .clear
```
Comment (sindresorhus) [inline Touch Bar Simulator/TouchBarWindow.swift]: I don't think this is needed as you already set the `borderWidth`.

File: Touch Bar Simulator/TouchBarWindow.swift [hunk 3]
```diff
@@ -295,6 +300,17 @@ final class TouchBarWindow: NSPanel {
 	func removeTitlebar() {
 		styleMask.remove(.titled)
 	}
+	
+	func roundCorners(_ corners: CACornerMask) {
+		self.contentView?.layer?.cornerRadius = 7.5
+		self.contentView?.layer?.maskedCorners = []
```
Comment (sindresorhus) [inline Touch Bar Simulator/TouchBarWindow.swift]: This is already set below.

File: Touch Bar Simulator/TouchBarWindow.swift [hunk 4]
```diff
@@ -295,6 +300,17 @@ final class TouchBarWindow: NSPanel {
 	func removeTitlebar() {
 		styleMask.remove(.titled)
 	}
+	
+	func roundCorners(_ corners: CACornerMask) {
+		self.contentView?.layer?.cornerRadius = 7.5
```
Comment (sindresorhus) [inline Touch Bar Simulator/TouchBarWindow.swift]: Maybe add a comment that the 7.5 number is to match the window corner radius.

File: Touch Bar Simulator/TouchBarWindow.swift [hunk 5]
```diff
@@ -295,6 +300,17 @@ final class TouchBarWindow: NSPanel {
...
+		self.contentView?.layer?.maskedCorners = []
+
+		self.contentView?.layer?.maskedCorners = corners
+		self.contentView?.layer?.borderWidth = 0.0
+		self.contentView?.layer?.borderColor = .clear
+		self.isOpaque = false
+		self.backgroundColor = .clear
```
Comment (sindresorhus) [inline Touch Bar Simulator/TouchBarWindow.swift]: Shouldn't this be set to the opposite here https://github.com/sindresorhus/touch-bar-simulator/pull/54/files#diff-e6aaf782da40568071255e5e366b109dR284?

File: Touch Bar Simulator/TouchBarWindow.swift [hunk 6]
```diff
@@ -11,8 +11,10 @@ final class TouchBarWindow: NSPanel {
 				window.addTitlebar()
 			case .dockedToTop:
 				window.removeTitlebar()
+				window.roundCorners([.layerMaxXMinYCorner, .layerMinXMinYCorner])
```
Comment (sindresorhus) [inline Touch Bar Simulator/TouchBarWindow.swift]: Add a couple of `CACornerMask` extensions to make this and the below call nicer. `.minY` and `.maxY`.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
@DeveloperBen Friendly bump :)

**Comment by sindresorhus** [comment]
This also needs to use corner radius on all 4 sides when docked with a padding: https://github.com/sindresorhus/touch-bar-simulator/pull/63

**Comment by ThinkFar** [comment]
Any update on this? What's missing? Why is this not merged?

**Pull request closed without merging**